src/operator_data.c

```diff
@@ -0,0 +1,391 @@
+  PetscFunctionBegin;
+  *boundary_data = (OperatorBoundaryData){0};
+  if (!rdy->lock.boundary_data) {
+    PetscCall(PetscCalloc1(rdy->num_boundaries, &rdy->lock.boundary_data));
+  }
+  PetscCheck(boundary.index >= 0 && boundary.index < rdy->num_boundaries, rdy->comm, PETSC_ERR_USER,
+             "Invalid boundary for boundary data (index: %" PetscInt_FMT ")", boundary.index);
+  PetscCheck(!rdy->lock.boundary_data[boundary.index], rdy->comm, PETSC_ERR_USER,
+             "Could not acquire lock on boundary data -- another entity has access");
+  rdy->lock.boundary_data[boundary.index] = boundary_data;
+  boundary_data->rdy                      = rdy;
+  boundary_data->boundary                 = boundary;
+  PetscCall(VecGetBlockSize(rdy->u_global, &boundary_data->num_components));
+
+  if (CeedEnabled(rdy)) {
+    // get the relevant boundary sub-operator
+    CeedOperator *sub_ops;
+    PetscCallCEED(CeedCompositeOperatorGetSubList(rdy->ceed.flux_operator, &sub_ops));
+    CeedOperator flux_op = sub_ops[1 + boundary_data->boundary.index];
+
+    // fetch the relevant vector
+    CeedOperatorField field;
+    PetscCallCEED(CeedOperatorGetFieldByName(flux_op, "q_dirichlet", &field));
+    PetscCallCEED(CeedOperatorFieldGetVector(field, &boundary_data->storage.ceed.vec));
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// sets boundary values for the given component on the boundary associated with
+// the given operator boundary data
+PetscErrorCode SetOperatorBoundaryValues(OperatorBoundaryData *boundary_data, PetscInt component, PetscReal *boundary_values) {
+  PetscFunctionBegin;
+
+  if (CeedEnabled(boundary_data->rdy)) {
+    // if this is the first update, get access to the vector's data
+    if (!boundary_data->storage.updated) {
+      PetscCallCEED(CeedVectorGetArray(boundary_data->storage.ceed.vec, CEED_MEM_HOST, &boundary_data->storage.ceed.data));
+      boundary_data->storage.updated = PETSC_TRUE;
+    }
+
+    // reshape for multicomponent access
+    CeedScalar(*values)[boundary_data->num_components];
+    *((CeedScalar **)&values) = boundary_data->storage.ceed.data;
+
+    // set the data
+    for (CeedInt e = 0; e < boundary_data->boundary.num_edges; ++e) {
+      values[e][component] = boundary_values[e];
+    }
+  } else {
+    // if this is the first update, get access to the vector's data
+    if (!boundary_data->storage.updated) {
+      PetscCall(VecGetArray(boundary_data->storage.petsc.vec, &boundary_data->storage.petsc.data));
+      boundary_data->storage.updated = PETSC_TRUE;
+    }
+
+    // set the data
+    PetscReal *u = boundary_data->storage.petsc.data;
+    for (PetscInt e = 0; e < boundary_data->boundary.num_edges; ++e) {
+      u[e * boundary_data->num_components + component] = boundary_values[e];
+    }
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+PetscErrorCode RestoreOperatorBoundaryData(RDy rdy, RDyBoundary boundary, OperatorBoundaryData *boundary_data) {
```

Comment:
A few functions are missing the comment describing the function, but it can be added in a future PR.

Reply:
Yeah, that sounds good.